src/server/stream_family.cc

```diff
@@ -1640,6 +1852,55 @@ void StreamFamily::XRead(CmdArgList args, ConnectionContext* cntx) {
   }
 }
 
+void XReadGeneric(CmdArgList args, bool read_group, ConnectionContext* cntx) {
+  auto opts = ParseReadArgsOrReply(args, read_group, cntx);
+  if (!opts) {
+    return;
+  }
+
+  vector<vector<GroupConsumerPair>> res_pairs(shard_set->size());
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    // if the command is `XREAD`, we don't need to get
+    // (group, consumer) pair.
+    if (!opts->read_group) {
+      return OpStatus::OK;
+    }
+    auto sid = shard->shard_id();
+    auto s_args = t->GetShardArgs(sid);
+    GroupConsumerPairOpts gc_opts = {opts->group_name, opts->consumer_name};
+
+    auto items_res = OpGetGroupConsumerPairs(s_args, t->GetOpArgs(shard), gc_opts);
+    res_pairs[sid] = items_res;
+    return OpStatus::OK;
+  };
+  cntx->transaction->Schedule();
+  cntx->transaction->Execute(std::move(cb), false);
+
+  // for xreadgroup, set the consumer and group fields of stream_ids items.
+  for (size_t i = 0; i < shard_set->size(); i++) {
```

Comment:
it's relevant only if  `opts->read_group` is set?

Reply:
yeah, in case of `xread`, the `res_pairs[i].size()` = 0

Reply:
in that case lets move it into the block `if (opts->read_group) {` above